### ISSUE TITLE
Welder refueling ignores the 'TankSafe' property, ignores whether reagents match, and ignores whether the welder has RefillableSolution at all

### ISSUE THREAD

**Issue by Elijahrane** [space-wizards/space-station-14#11487]
## Description

Really hampers any welders that violate any of these assumptions because it's overly hardcoded. Just noting it here for future tool refactors

**Comment by Mervill** [comment]
I think the problem with `TankSafe` is that nothing's actually marked as `TankSafe` except lighters(?). But Welding Tanks have the correct `DamageOnToolInteract` component added. Which welders should be `TankSafe`? 
Relevant function: `DamageOnToolInteractSystem::OnInteracted()`

Reagent matching is sorta implied by `ReagentTankComponent` but that component could probably be removed (only used in this one spot) in favor of generic reagent checking. This is where we'd ensure the welder has `RefillableSolution`. 
Relevant function: `ToolSystem::OnWelderAfterInteract()`

Explosion via welding for the refueling tank `WeldingFuelTank` is handled in `SolutionExplosionBehavior` (on the parent prototype `StorageTank`). If the object takes heat damage and the solution isn't empty, the tank will explode, otherwise not.
Relevant class: `SolutionExplosionBehavior`

**Comment by Elijahrane** [comment]
Sorry, might be unclear about the mismatch. What I meant is that if you have a welder that uses a reagent that is not welding fuel, the tank will fill it with welding fuel anyway. This permanently reduces the capacity because you can't remove the welding fuel through burning or anything

**Comment by Mervill** [comment]
The empty tank does not get destroyed by an explosion specifically (check fails because tank is empty). The tank takes heat damage from the lit Welder and will either explode or simply be destroyed. It dosen't really have a destroyed state right now which makes it look kind of janky. Should trying to refuel from an empty tank simply not do anything?

Am I crazy or does `RefuelableSolutionComponent` not exist?

**Comment by mirrorcult** [comment]
> Am I crazy or does `RefuelableSolutionComponent` not exist? Or are you suggesting we create that component to deal with refueling?

i dont believe it exists right now, it seems like a reasonable thing to add to separate welder a bit

**Comment by Elijahrane** [comment]
I meant ```RefillableSolutionComponent```

**Comment by Mervill** [comment]
ah ok sorry I that was confusing me 😣

**Comment by Elijahrane** [comment]
yeah my writing on this issue has been pretty unclear but it's accurate now I think

**Comment by IProduceWidgets** [comment]
Replicated that you can (theoretically, afaik there's no instance of this upstream by default) refill a welder from an entity with the ReagentTankComponent with TankType set to Fuel with whatever the reagent the tank's solution container holds. This then prevents a welder from burning since it doesn't have the correct fuel.

The discussion above concludes that whether the solution is used to refill the welder should be determined by data *on the welder* not the ReagentTank. 

I'll note here that the WelderComponent already has a field on the welder entity to specify the FuelReagent, and could probably be used for this.


An alternative solution is to just let welders be opened and spilled.

**Comment by deltanedas** [comment]
welder needs rewriting anyway
